pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package operationexecutor implements interfaces that enable execution of
+// register and unregister operations with a
+// nestedpendingoperations so that more than one operation is never triggered
+// on the same plugin for the same pod.
+package operationexecutor
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
+)
+
+var _ OperationGenerator = &operationGenerator{}
+
+type operationGenerator struct {
+
+	// recorder is used to record events in the API server
+	recorder record.EventRecorder
+}
+
+// NewOperationGenerator is returns instance of operationGenerator
+func NewOperationGenerator(recorder record.EventRecorder) OperationGenerator {
+
+	return &operationGenerator{
+		recorder: recorder,
+	}
+}
+
+// OperationGenerator interface that extracts out the functions from operation_executor to make it dependency injectable
+type OperationGenerator interface {
+	// Generates the RegisterPlugin function needed to perform the registration of a plugin
+	GenerateRegisterPluginFunc(
+		pluginToRegister cache.PluginInfo,
+		pluginHandler cache.PluginHandler,
+		actualStateOfWorldUpdater ActualStateOfWorldUpdater) (func() error, error)
```

Comment:
Drop the `error` from the return parameters of the `Generate...` methods. When there is an option to return an error, people will start adding validation to the `Generate...` methods. And the problem with that is the reconciler is dumb, it will just call the same method again, and that will fail, and that will cause the logs to be filled with the same error again and again.

Instead we should encourage people to do all the work, including the validation inside the function that is returned, therefore if validation fails, the `goroutinemap` framework handles exponential back off. (See https://github.com/kubernetes/kubernetes/pull/71581/files for an example of this change.)